File: api/Security.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Security 对应 Security.py。
type Security struct {
	*session.ApiGroup
}

// NewSecurity 创建 Security API 分组。
func NewSecurity(s *session.Session) *Security {
	return &Security{ApiGroup: session.NewApiGroup(s)}
}

// Bridgemode 对应 security/bridgemode。
func (s *Security) Bridgemode() (map[string]interface{}, error) {
	return s.S.Get("security/bridgemode", nil, "api")
}

// GetFirewallSwitch 获取防火墙开关。对应 security/firewall-switch。
func (s *Security) GetFirewallSwitch() (map[string]interface{}, error) {
	return s.S.Get("security/firewall-switch", nil, "api")
}

// SetFirewallSwitch 设置防火墙开关。对应 security/firewall-switch (post_set)。
func (s *Security) SetFirewallSwitch(
	firewall bool,
	ipFilter bool,
	wanPingFilter bool,
	urlFilter bool,
	macFilter bool,
) (interface{}, error) {
	return s.S.PostSet("security/firewall-switch", session.O(
		"FirewallMainSwitch", boolToInt(firewall),
		"FirewallIPFilterSwitch", boolToInt(ipFilter),
		"FirewallWanPortPingSwitch", boolToInt(wanPingFilter),
		"firewallurlfilterswitch", boolToInt(urlFilter),
		"firewallmacfilterswitch", boolToInt(macFilter),
	), false, "api", false, false)
}

// MacFilter 对应 security/mac-filter。
func (s *Security) MacFilter() (map[string]interface{}, error) {
	return s.S.Get("security/mac-filter", nil, "api")
}

// LanIpFilter 对应 security/lan-ip-filter。
func (s *Security) LanIpFilter() (map[string]interface{}, error) {
	return s.S.Get("security/lan-ip-filter", nil, "api")
}

// VirtualServers 对应 security/virtual-servers。
func (s *Security) VirtualServers() (map[string]interface{}, error) {
	return s.S.Get("security/virtual-servers", nil, "api")
}

// UrlFilter 对应 security/url-filter。
func (s *Security) UrlFilter() (map[string]interface{}, error) {
	return s.S.Get("security/url-filter", nil, "api")
}

// SetUrlFilter 设置 URL 过滤。对应 security/url-filter (post_set)。
// urlfilters 结构与 UrlFilter 返回值相同。
func (s *Security) SetUrlFilter(urlfilters map[string]interface{}) (interface{}, error) {
	return s.S.PostSet("security/url-filter", urlfilters, false, "api", false, false)
}

// Upnp 对应 security/upnp。
func (s *Security) Upnp() (map[string]interface{}, error) {
	return s.S.Get("security/upnp", nil, "api")
}

// SetUpnp 设置 UPnP。对应 security/upnp (post_set)。
func (s *Security) SetUpnp(enabled bool) (interface{}, error) {
	return s.S.PostSet("security/upnp", map[string]interface{}{
		"UpnpStatus": boolToInt(enabled),
	}, false, "api", false, false)
}

// Dmz 对应 security/dmz。
func (s *Security) Dmz() (map[string]interface{}, error) {
	return s.S.Get("security/dmz", nil, "api")
}

// SetDmz 设置 DMZ。对应 security/dmz (post_set)。
func (s *Security) SetDmz(enabled bool, ipAddress string) (interface{}, error) {
	return s.S.PostSet("security/dmz", session.O(
		"DmzStatus", boolToInt(enabled),
		"DmzIPAddress", ipAddress,
	), false, "api", false, false)
}

// Sip 对应 security/sip。
func (s *Security) Sip() (map[string]interface{}, error) {
	return s.S.Get("security/sip", nil, "api")
}

// SetSip 设置 SIP。对应 security/sip (post_set)。
func (s *Security) SetSip(enabled bool, port int) (interface{}, error) {
	return s.S.PostSet("security/sip", session.O(
		"SipStatus", boolToInt(enabled),
		"SipPort", port,
	), false, "api", false, false)
}

// FeatureSwitch 对应 security/feature-switch。
func (s *Security) FeatureSwitch() (map[string]interface{}, error) {
	return s.S.Get("security/feature-switch", nil, "api")
}

// Nat 对应 security/nat。
func (s *Security) Nat() (map[string]interface{}, error) {
	return s.S.Get("security/nat", nil, "api")
}

// SpecialApplications 对应 security/special-applications。
func (s *Security) SpecialApplications() (map[string]interface{}, error) {
	return s.S.Get("security/special-applications", nil, "api")
}

// WhiteLanIpFilter 对应 security/white-lan-ip-filter。
func (s *Security) WhiteLanIpFilter() (map[string]interface{}, error) {
	return s.S.Get("security/white-lan-ip-filter", nil, "api")
}

// WhiteUrlFilter 对应 security/white-url-filter。
func (s *Security) WhiteUrlFilter() (map[string]interface{}, error) {
	return s.S.Get("security/white-url-filter", nil, "api")
}

// SetWhiteUrlFilter 设置白名单 URL 过滤。对应 security/white-url-filter (post_set)。
// urlfilters 结构与 WhiteUrlFilter 返回值相同。
func (s *Security) SetWhiteUrlFilter(urlfilters map[string]interface{}) (interface{}, error) {
	return s.S.PostSet("security/white-url-filter", urlfilters, false, "api", false, false)
}

// Acls 对应 security/acls。
func (s *Security) Acls() (map[string]interface{}, error) {
	return s.S.Get("security/acls", nil, "api")
}

// Acl 对应 security/acl。
func (s *Security) Acl() (map[string]interface{}, error) {
	return s.S.Get("security/acl", nil, "api")
}

File: src/api/security.rs
//! Security API group (`api/Security.py`).
//!
//! Firewall, MAC/IP/URL filtering, NAT and port-forwarding.

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_set};

/// Security API group.
pub struct Security<'a> {
    conn: &'a Connection,
}

impl<'a> Security<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Security { conn }
    }

    /// `security/bridgemode`. Bridge mode state.
    pub fn bridgemode(&self) -> Result<Value> {
        get_value(self.conn, "api/security/bridgemode")
    }

    /// `security/firewall-switch`. Firewall switch state.
    pub fn get_firewall_switch(&self) -> Result<Value> {
        get_value(self.conn, "api/security/firewall-switch")
    }

    /// Toggle the individual firewall filters.
    pub fn set_firewall_switch(
        &self,
        firewall: bool,
        ip_filter: bool,
        wan_ping_filter: bool,
        url_filter: bool,
        mac_filter: bool,
    ) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "FirewallMainSwitch",
                if firewall {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            (
                "FirewallIPFilterSwitch",
                if ip_filter {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            (
                "FirewallWanPortPingSwitch",
                if wan_ping_filter {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            (
                "firewallurlfilterswitch",
                if url_filter {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            (
                "firewallmacfilterswitch",
                if mac_filter {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
        ]);
        post_set(self.conn, "api/security/firewall-switch", &body)
    }

    /// `security/mac-filter`. MAC filter state.
    pub fn mac_filter(&self) -> Result<Value> {
        get_value(self.conn, "api/security/mac-filter")
    }

    /// `security/lan-ip-filter`. LAN IP filter state.
    pub fn lan_ip_filter(&self) -> Result<Value> {
        get_value(self.conn, "api/security/lan-ip-filter")
    }

    /// `security/virtual-servers`. Port-forward (virtual server) rules.
    pub fn virtual_servers(&self) -> Result<Value> {
        get_value(self.conn, "api/security/virtual-servers")
    }

    /// `security/url-filter`. URL filter rules.
    pub fn url_filter(&self) -> Result<Value> {
        get_value(self.conn, "api/security/url-filter")
    }

    /// Replace the URL filter rules.
    ///
    /// * `urlfilters` — same structure as [`Security::url_filter`] returns.
    pub fn set_url_filter(&self, urlfilters: &XmlMap) -> Result<String> {
        post_set(self.conn, "api/security/url-filter", urlfilters)
    }

    /// `security/upnp`. UPnP state.
    pub fn upnp(&self) -> Result<Value> {
        get_value(self.conn, "api/security/upnp")
    }

    /// Toggle UPnP.
    pub fn set_upnp(&self, enabled: bool) -> Result<String> {
        let body: XmlMap = map_of([(
            "UpnpStatus",
            if enabled {
                "1".to_string()
            } else {
                "0".to_string()
            },
        )]);
        post_set(self.conn, "api/security/upnp", &body)
    }

    /// `security/dmz`. DMZ state.
    pub fn dmz(&self) -> Result<Value> {
        get_value(self.conn, "api/security/dmz")
    }

    /// Enable/disable DMZ and set the exposed host's IP.
    pub fn set_dmz(&self, enabled: bool, ip_address: &str) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "DmzStatus",
                if enabled {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            ("DmzIPAddress", ip_address.to_string()),
        ]);
        post_set(self.conn, "api/security/dmz", &body)
    }

    /// `security/sip`. SIP ALG state.
    pub fn sip(&self) -> Result<Value> {
        get_value(self.conn, "api/security/sip")
    }

    /// Enable/disable the SIP application-layer gateway and set its port.
    pub fn set_sip(&self, enabled: bool, port: i64) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "SipStatus",
                if enabled {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            ("SipPort", port.to_string()),
        ]);
        post_set(self.conn, "api/security/sip", &body)
    }

    /// `security/feature-switch`. Feature switch state.
    pub fn feature_switch(&self) -> Result<Value> {
        get_value(self.conn, "api/security/feature-switch")
    }

    /// `security/nat`. NAT state.
    pub fn nat(&self) -> Result<Value> {
        get_value(self.conn, "api/security/nat")
    }

    /// `security/special-applications`. Special application rules.
    pub fn special_applications(&self) -> Result<Value> {
        get_value(self.conn, "api/security/special-applications")
    }

    /// `security/white-lan-ip-filter`. Whitelist LAN IP filter state.
    pub fn white_lan_ip_filter(&self) -> Result<Value> {
        get_value(self.conn, "api/security/white-lan-ip-filter")
    }

    /// `security/white-url-filter`. Whitelist URL filter rules.
    pub fn white_url_filter(&self) -> Result<Value> {
        get_value(self.conn, "api/security/white-url-filter")
    }

    /// Replace the whitelist URL filter rules.
    ///
    /// * `urlfilters` — same structure as [`Security::white_url_filter`] returns.
    pub fn set_white_url_filter(&self, urlfilters: &XmlMap) -> Result<String> {
        post_set(self.conn, "api/security/white-url-filter", urlfilters)
    }

    /// `security/acls` (reverse engineered, likely unused).
    pub fn acls(&self) -> Result<Value> {
        get_value(self.conn, "api/security/acls")
    }

    /// `security/acl`. ACL rule.
    pub fn acl(&self) -> Result<Value> {
        get_value(self.conn, "api/security/acl")
    }
}
